xml.c

```diff
@@ -23,54 +23,54 @@ int stepInto(xmlTextReaderPtr reader) {
 	int result;
 	do {
 		result = xmlTextReaderRead(reader);
-	} while (result
+	} while (result > 0
 			&& (xmlTextReaderNodeType(reader) == XML_READER_TYPE_SIGNIFICANT_WHITESPACE
 					|| xmlTextReaderNodeType(reader) == XML_READER_TYPE_COMMENT
 					|| xmlTextReaderNodeType(reader) == XML_READER_TYPE_END_ELEMENT));
-	return result;
+	return result > 0;
 }
 
 int stepOver(xmlTextReaderPtr reader) {
 	int depth = xmlTextReaderDepth(reader);
 	int result;
 	do {
 		result = xmlTextReaderRead(reader);
-	} while (result && xmlTextReaderDepth(reader) > depth);
-	while (result
+	} while (result > 0 && xmlTextReaderDepth(reader) > depth);
+	while (result > 0
 			&& (xmlTextReaderNodeType(reader) == XML_READER_TYPE_SIGNIFICANT_WHITESPACE
 					|| xmlTextReaderNodeType(reader) == XML_READER_TYPE_COMMENT
 					|| xmlTextReaderNodeType(reader) == XML_READER_TYPE_END_ELEMENT)) {
 		result = xmlTextReaderRead(reader);
 	}
-	return result;
+	return result > 0;
 }
 
 int stepOut(xmlTextReaderPtr reader) {
 	int depth = xmlTextReaderDepth(reader) - 1;
 	int result;
 	do {
 		result = xmlTextReaderRead(reader);
-	} while (result && xmlTextReaderDepth(reader) > depth);
-	while (result
+	} while (result > 0 && xmlTextReaderDepth(reader) > depth);
+	while (result > 0
 			&& (xmlTextReaderNodeType(reader) == XML_READER_TYPE_SIGNIFICANT_WHITESPACE
 					|| xmlTextReaderNodeType(reader) == XML_READER_TYPE_COMMENT
 					|| xmlTextReaderNodeType(reader) == XML_READER_TYPE_END_ELEMENT)) {
 		result = xmlTextReaderRead(reader);
 	}
-	return result;
+	return result > 0;
 }
 
 int stepOverText(xmlTextReaderPtr reader, const char ** text) {
 	int depth = xmlTextReaderDepth(reader);
 	int result = stepInto(reader);
 	*text = NULL;
-	if (result && xmlTextReaderDepth(reader) > depth) {
+	if (result > 0 && xmlTextReaderDepth(reader) > depth) {
 		if (xmlTextReaderNodeType(reader) == XML_READER_TYPE_TEXT) {
 			*text = xmlTextReaderValue(reader);
 		}
 		result = stepOut(reader);
 	}
-	return result;
+	return result > 0;
```

Comment:
I don't know why I missed error handling in this code, that's a bit of a mystery.

However the fix shouldn't deviate from the [return codes](http://www.xmlsoft.org/html/libxml-xmlreader.html#xmlTextReaderRead) in libxml2:
> Returns:	1 if the node was read successfully, 0 if there is no more nodes to read, or -1 in case of error

In other words `0` means EOF and `-1` mean error.  For all `return > 0` fixes the immediate problem of other calls failing to check error, the fix should really be to simply return `result` and fix those other lines.  That lets the code `exit(1)` at an appropriate point.

Reply:
The existing code handles it ok, it'll skip the problem section and continue parsing.  If this returns -1, then all the code using this needs to be updated too, so it detects errors and handles them specially.

Reply:
Yeah. Sadly that's what it needs. :-(
I'll try to assess how much damage this will do to the rest of the code.  It might not be so bad.

Reply:
I've tested a bunch of different errors and it's decent what happens.  It doesn't abort, but the parts that aren't bad work ok and it starts.  IMHO, it would be better to abort on any error and produce a clear message about it.